### PR TITLE
[OSDOCS-9373]: Azure CPMS failure domain misc updates (prework)

### PR DESCRIPTION
**Pull request by jeana-redhat**
Version(s):
4.15+

Issue:
[OSDOCS-9373](https://issues.redhat.com//browse/OSDOCS-9373)

Link to docs preview:
* [Sample AWS provider specification](https://63227--ocpdocs-pr.netlify.app/openshift-enterprise/latest/machine_management/control_plane_machine_management/cpmso-configuration#cpmso-yaml-provider-spec-aws_cpmso-configuration)
* [Sample GCP failure domain configuration](https://63227--ocpdocs-pr.netlify.app/openshift-enterprise/latest/machine_management/control_plane_machine_management/cpmso-configuration#cpmso-yaml-failure-domain-gcp_cpmso-configuration) (this section was mistakenly commented out from the page - no changes)
* [Sample Azure provider specification](https://63227--ocpdocs-pr.netlify.app/openshift-enterprise/latest/machine_management/control_plane_machine_management/cpmso-configuration#cpmso-yaml-provider-spec-azure_cpmso-configuration)
* [Sample Azure failure domain configuration](https://63227--ocpdocs-pr.netlify.app/openshift-enterprise/latest/machine_management/control_plane_machine_management/cpmso-configuration#cpmso-yaml-failure-domain-azure_cpmso-configuration)
* [Sample Nutanix provider specification](https://63227--ocpdocs-pr.netlify.app/openshift-enterprise/latest/machine_management/control_plane_machine_management/cpmso-configuration#cpmso-yaml-provider-spec-nutanix_cpmso-configuration)

QE review:
- [x] QE has approved this change.

Additional information:
This was originally for the multi-subnet feature, but that is no longer in the release. However, some edits from that effort and some related changes should still be merged for 4.15

### CODE REVIEW

**Comment by jeana-redhat** [comment]
@RadekManak can you PTAL at this PR when you have some time?

File: modules/cpmso-yaml-failure-domain-azure.adoc
Comment (RadekManak) [inline modules/cpmso-yaml-failure-domain-azure.adoc:30]: This is techpreview in 4.14 because the CPMS work did not merge. It will likely be GA in 4.15 and default on new clusters.

File: modules/cpmso-yaml-failure-domain-azure.adoc
Comment (RadekManak) [inline modules/cpmso-yaml-failure-domain-azure.adoc:26]: I am not entirely happy how we are just telling customers to not touch this. Ideally, we would explain the differences between the two network stacks and provide steps to convert from single subnet cluster to multi subnet cluster. 

But here for explaining the subnet field it is good enough, I guess.

**Review by RadekManak — CHANGES_REQUESTED** [review]
This feature did not make it to 4.14. It should not be merged yet.

**Comment by jeana-redhat** [comment]
Thanks! I missed that this slipped. Closing and removing labels, but will keep the links around for 4.15. Thank you!

File: modules/cpmso-yaml-failure-domain-azure.adoc
Comment (jeana-redhat) [inline modules/cpmso-yaml-failure-domain-azure.adoc:13]: @RadekManak can you confirm which versions of OCP this format (including a `subnet` field, even though it's just one subnet) impacts? I think only 4.15+, but want to make sure.

**Comment by RadekManak** [comment]
Above in providerSpec example [9] Specifies the subnet for the control plane. Should be replaced with "This parameter is configured in the failure domain, and is shown with an empty value here. If a value specified for this parameter differs from the value in the failure domain, the Operator overwrites it with the value in the failure domain."

There has also been a slight change on how failuredomains are applied. This logic is the same across all platforms, not just azure. 
You can specify a value in providerSpec as a default value. If you specify that field for a given failureDomain it will be replaced. If you don't specify, the default value from providerSpec will be kept. 

This is to support clusters updating from old versions where they don't have subnet set in failureDomain, but in the providerSpec instead. Clusters updated from 4.14 will have:

providerSpec:
  .....
  subnet: subnetname
  ....

failuredomain:
  - zone: "1" 
  - zone: "2" 
  - zone: "3" 

Not sure if that needs to be mentioned in docs.

**Comment by jeana-redhat** [comment]
> Above in providerSpec example [9] Specifies the subnet for the control plane. Should be replaced with "This parameter is configured in the failure domain, and is shown with an empty value here. If a value specified for this parameter differs from the value in the failure domain, the Operator overwrites it with the value in the failure domain."

Excellent catch

> There has also been a slight change on how failuredomains are applied. This logic is the same across all platforms, not just azure. You can specify a value in providerSpec as a default value. If you specify that field for a given failureDomain it will be replaced. If you don't specify, the default value from providerSpec will be kept.
> 
> This is to support clusters updating from old versions where they don't have subnet set in failureDomain, but in the providerSpec instead. Clusters updated from 4.14 will have:
> 
> providerSpec: ..... subnet: subnetname ....
> 
> failuredomain:
> 
>     * zone: "1"
> 
>     * zone: "2"
> 
>     * zone: "3"
> 
> 
> Not sure if that needs to be mentioned in docs.

Yes, I think we probably ought to explain that. I am going to push a change with the other change today, and think more about the second part of this tomorrow.

**Comment by jeana-redhat** [comment]
@RadekManak most recent commit is my attempt to explain the 4.15+ behavior. See what you think. 

I have made some notes ("Check for 4.15 FD behavior") on non-subnet fields that are also said to be configured in the FC (such as `zone`, etc.). Not sure if this applies there as well.

Updated Nutanix to show an empty subnet as well, since they are adding FD support ([my PR](https://github.com/openshift/openshift-docs/pull/70056) for that, and [Mike's related one](https://github.com/openshift/openshift-docs/pull/69618) for further reference)

**Review by RadekManak — COMMENTED** [review]
The failure domains are not applied the same across all platforms. AWS, GCP, Azure, Openstack. Work fine with some fields in providerSpec and omitted from failure domains.

And others only allow user changes in failure domains and overwrite values.

File: modules/cpmso-yaml-provider-spec-openstack.adoc
Comment (jeana-redhat) [reply modules/cpmso-yaml-provider-spec-openstack.adoc:48]: > Not exactly sure what you want to know here.

Mostly just checking to make sure nothing looks weird/outdated :slightly_smiling_face: 

Looks like current docs cover `AvailabilityZone` and `RootVolume.AvailabilityZone` in the sample failure domain YAML. I don't see `RootVolume.VolumeType` anywhere, but I'm also a bit hesitant to dig in here since OSP is outside the scope of Cloud team for the most part. I will drop a note to the OSP writer though just for his awareness.

File: modules/cpmso-yaml-provider-spec-vsphere.adoc
Comment (jeana-redhat) [reply modules/cpmso-yaml-provider-spec-vsphere.adoc:54]: Do you have a Jira item handy for this implementation?

File: modules/cpmso-yaml-provider-spec-nutanix.adoc
Comment (jeana-redhat) [reply modules/cpmso-yaml-provider-spec-nutanix.adoc:74]: Ack - will rewrite this one to try to get that all stated clearly

File: modules/cpmso-yaml-provider-spec-azure.adoc
Comment (jeana-redhat) [reply modules/cpmso-yaml-provider-spec-azure.adoc:71]: Ok, I am editing to try to explain that. I will write it up for but subnet and zone just to make sure it makes sense, but remove the subnet part before merge

Comment (jeana-redhat) [reply modules/cpmso-yaml-provider-spec-azure.adoc:71]: Actually, I am not 100% sue of one part of this, but it will be complicated to discuss in GH. I'll message you about it :)

File: modules/cpmso-yaml-provider-spec-aws.adoc
Comment (jeana-redhat) [reply modules/cpmso-yaml-provider-spec-aws.adoc:76]: I'm not sure what you mean by "shared values", we can chat about this later. Going to push an update with the rest of my edits.

File: modules/cpmso-yaml-provider-spec-nutanix.adoc
Comment (RadekManak) [reply modules/cpmso-yaml-provider-spec-nutanix.adoc:87]: Same here for when failure domain is specified.
If you specify this value in the provider specification, the Operator **ignores** it.

File: modules/cpmso-yaml-provider-spec-azure.adoc
Comment (jeana-redhat) [reply modules/cpmso-yaml-provider-spec-azure.adoc:74]: going to put "ignores" here so it's the same as the others, lmk if that's wrong though :)

File: modules/cpmso-yaml-provider-spec-vsphere.adoc
Comment (jeana-redhat) [reply modules/cpmso-yaml-provider-spec-vsphere.adoc:54]: thanks! will look into that. Pushing other changes for now.

Comment (jeana-redhat) [reply modules/cpmso-yaml-provider-spec-vsphere.adoc:54]: Ok, looks like another writer covered this one. I will remove that note and then I think this can be sent for QE. Let me know if you think I have missed anything in our conversation on this PR :)

**Comment by jeana-redhat** [comment]
@miyadav PTAL at these refinements in the CPMS config docs :bow:

**Comment by miyadav** [comment]
Thanks @jeana-redhat sorry for the delay , Is it ok if we add the exact place where to add this failureDomain for the end user of docs , if there is already a reference I might have missed which conveys where to add  exactly in the yaml , let me know , then we don't need below explicitly .example below ( same for all providers), Other than this LGTM .
```
failureDomains:
  azure:
  - zone: "1" 
  - zone: "2"
  - zone: "3"
  platform: Azure 
```
  
  to 
```
  `spec:
     .
     .
         template:
            .
            .
             failureDomains:
                 azure:
                  - zone: "1" 
                  - zone: "2"
                  - zone: "3"
                   platform: Azure `
```

**Comment by jeana-redhat** [comment]
@miyadav thanks for the feedback! 

My initial strategy on this was to include the reference for `providerSpec` and `failureDomains` here: [Sample YAML for a control plane machine set custom resource](https://63227--ocpdocs-pr.netlify.app/openshift-enterprise/latest/machine_management/control_plane_machine_management/cpmso-configuration#cpmso-yaml-sample-cr_cpmso-configuration).

However, given how long this page is, I think your suggestion makes more sense and it also aligns with docs best practices. I will add the hierarchy for orientation in all the `providerSpec` and `failureDomains` snippets.

**Comment by jeana-redhat** [comment]
@miyadav PTAL at the updated YAML snippets to make sure none of the added context introduced any errors :bow: I have left this change as a second commit, hope that helps!

File: modules/cpmso-yaml-provider-spec-aws.adoc
Comment (skopacz1) [inline modules/cpmso-yaml-provider-spec-aws.adoc:81]: Should you specify which Operator is performing this action? Maybe it's clear to users embedded in this workflow/content, but it might not hurt to also be more specific if possible.

File: modules/cpmso-yaml-provider-spec-aws.adoc
Comment (skopacz1) [inline modules/cpmso-yaml-provider-spec-aws.adoc:87]: Same comment about potentially specifying the Operator.

File: modules/cpmso-yaml-provider-spec-azure.adoc
Comment (skopacz1) [inline modules/cpmso-yaml-provider-spec-azure.adoc:84]: Should the Operator be specified here?

File: modules/cpmso-yaml-provider-spec-vsphere.adoc
Comment (skopacz1) [inline modules/cpmso-yaml-provider-spec-vsphere.adoc:1]: Out of scope and not necessary for this PR, but it might be worth adding "VMware" in front of "vSphere" on line 7 to be consistent with the other vSphere headings that surround this section.

**Review by skopacz1 — COMMENTED** [review]
Left a few comments but otherwise LGTM! I'm still relatively new to reviewing, so please wait until @adellape approves my review before implementing any feedback. Thanks!

File: modules/cpmso-yaml-provider-spec-aws.adoc
Comment (adellape) [inline modules/cpmso-yaml-provider-spec-aws.adoc:81]: Comma possibly not needed here? 🤔 
```suggestion
<10> This parameter is configured in the failure domain and is shown with an empty value here. If a value specified for this parameter differs from the value in the failure domain, the Operator overwrites it with the value in the failure domain.
```

**Review by adellape — COMMENTED** [review]
@skopacz1 @jeana-redhat Peer review comments LGTM 👍

File: modules/cpmso-yaml-failure-domain-azure.adoc
Comment (jeana-redhat) [reply modules/cpmso-yaml-failure-domain-azure.adoc:38]: Better?
```suggestion
If the cluster is configured to use a single zone for all failure domains, the `zone` parameter is configured in the provider specification instead of in the failure domain configuration.
```

Comment (skopacz1) [reply modules/cpmso-yaml-failure-domain-azure.adoc]: Yes, thanks!

**Comment by jeana-redhat** [comment]
@kalexand-rh got QE verification of the last update here, squashed commits. This is ready for merge if you approve.

**Review by kalexand-rh — APPROVED** [review]
This is approved to merge during freeze.

**Pull request merged**
